test/distributed/org/apache/cassandra/distributed/test/NativeTransportEncryptionOptionsTest.java

```diff
@@ -219,4 +234,118 @@ public void nodeMustNotStartWithNonExistantCiphersTest() throws Throwable
+                                                     .trustManager(createTrustManagerFactory("test/conf/cassandra_ssl_test.truststore", "cassandra"))
+                                                     .build();
+            final SSLOptions sslOptions = socketChannel -> sslContext.newHandler(socketChannel.alloc());
+            com.datastax.driver.core.Cluster driverCluster = com.datastax.driver.core.Cluster.builder()
+                                                                                             .addContactPoint(address.getHostAddress())
+                                                                                             .withSSL(sslOptions)
+                                                                                             .build();
+            expectedException.expect(NoHostAvailableException.class);
+            driverCluster.connect();
+        }
+    }
+
+    @Test
+    public void testEndpointVerificationEnabledWithIPInSan() throws Throwable
+    {
+        // When required_endpoint_verification is set to true, client certificate Ip/hostname should be validated
+        // The certificate in cassandra_ssl_test_outbound.keystore have IP/hostname emebeded, so when
+        // require_endpoint_verification is true, the connection should be established
+        try (Cluster cluster = builder().withNodes(1).withConfig(c -> {
+            c.with(Feature.NATIVE_PROTOCOL);
+            c.set("client_encryption_options",
+                  ImmutableMap.builder().putAll(validKeystore)
+                              .put("enabled", true)
+                              .put("require_client_auth", true)
+                              .put("require_endpoint_verification", true)
+                              .build());
+        }).start())
+        {
+            InetAddress address = cluster.get(1).config().broadcastAddress().getAddress();
+            SslContext sslContext = SslContextBuilder.forClient()
+                                                     .keyManager(createKeyManagerFactory("test/conf/cassandra_ssl_test_endpoint_verify.keystore", "cassandra"))
+                                                     .trustManager(createTrustManagerFactory("test/conf/cassandra_ssl_test.truststore", "cassandra"))
+                                                     .build();
+            final SSLOptions sslOptions = socketChannel -> sslContext.newHandler(socketChannel.alloc());
+            com.datastax.driver.core.Cluster driverCluster = com.datastax.driver.core.Cluster.builder()
+                                                                                             .addContactPoint(address.getHostAddress())
+                                                                                             .withSSL(sslOptions)
+                                                                                             .build();
+            driverCluster.connect();
+        }
+    }
```

Comment:
nit: there are fair amount of code duplication in the 3 new tests. Can you refactor to reduce the dups? For example, 

```java
    private void testEndpointVerification(boolean requireEndpointVerification, boolean ipInSAN) throws Throwable
    {
        try (Cluster cluster = builder().withNodes(1).withConfig(c -> {
            c.with(Feature.NATIVE_PROTOCOL);
            c.set("client_encryption_options",
                  ImmutableMap.builder().putAll(validKeystore)
                              .put("enabled", true)
                              .put("require_client_auth", true)
                              .put("require_endpoint_verification", requireEndpointVerification)
                              .build());
        }).start())
        {
            InetAddress address = cluster.get(1).config().broadcastAddress().getAddress();
            SslContextBuilder sslContextBuilder = SslContextBuilder.forClient();
            if (ipInSAN)
                sslContextBuilder.keyManager(createKeyManagerFactory("test/conf/cassandra_ssl_test_endpoint_verify.keystore", "cassandra"));
            else
                sslContextBuilder.keyManager(createKeyManagerFactory("test/conf/cassandra_ssl_test_outbound.keystore", "cassandra"));

            SslContext sslContext = sslContextBuilder.trustManager(createTrustManagerFactory("test/conf/cassandra_ssl_test.truststore", "cassandra"))
                                                     .build();
            final SSLOptions sslOptions = socketChannel -> sslContext.newHandler(socketChannel.alloc());
            com.datastax.driver.core.Cluster driverCluster = com.datastax.driver.core.Cluster.builder()
                                                                                             .addContactPoint(address.getHostAddress())
                                                                                             .withSSL(sslOptions)
                                                                                             .build();

            if (!ipInSAN)
            {
                expectedException.expect(NoHostAvailableException.class);
            }

            driverCluster.connect();
        }
    }
```